### PR TITLE
<Input/> - refactor component

### PR DESCRIPTION
**Pull request by EgoziE**
A complete re-arrange for Input component.
I also removed irrelevant code selecting inner input text that was made in 2017 due to an old bug in Opera browser that was already fixed.

### CODE REVIEW

**Comment by WixBuildServer** [comment]
### 🤖 Hi, I'm #Janet!

  A deploy preview of your PR was successfully released 🖼.

  You can find it under one of the following URLs:
  * https://preview.wix-style-react.com/pr_6409

File: src/Input/Input.js [hunk 1]
```diff
@@ -67,21 +67,192 @@ class Input extends Component {
...
+  _onChange = (event) => {
+    const { onChange } = this.props;
+
+    if (this._isValidInput(event.target.value)) {
+      onChange && onChange(event);
+    }
+  };
+
+  _onKeyPress = (e) => {
+    if (!this._isValidInput(e.key)) {
```
Comment (gendeld) [inline src/Input/Input.js:172]: For number this is ok, but maybe if we introduce more complex types in the future checking only last char isn't enough.
Maybe check
```
`${value}${e.key}`
```

File: src/Input/Input.js [hunk 2]
```diff
@@ -67,21 +67,192 @@ class Input extends Component {
...
+      e.preventDefault();
+    }
+  };
+
+  _triggerOnChangeHandlerOnClearEvent = (event) => {
```
Comment (gendeld) [inline src/Input/Input.js:177]: This isn't being used anywhere. Can we remove it?

File: src/Input/Input.js (hunk 1 above)
Comment (EgoziE) [reply src/Input/Input.js:172]: You're right!
I was trying to create a more "generic" validation function in order to enable future need (for example a credit card validation)

File: src/Input/Input.js [hunk 3]
```diff
@@ -67,21 +67,192 @@ class Input extends Component {
...
+      inputProps,
+    );
+  };
+
+  /**
+   * Sets focus on the input element
+   * @param {FocusOptions} options
+   */
+  focus = (options = {}) => {
+    this._onFocus();
```
Comment (gendeld) [inline src/Input/Input.js:222]: Aren't we already passing _onFocus as `onFocus` prop for input?
Won't this cause _onFocus to be triggered twice?

File: src/Input/Input.js [hunk 4]
```diff
@@ -67,21 +67,192 @@ class Input extends Component {
...
+
+    if (!this._isControlled) {
+      this.input.value = '';
+    }
+
+    onClear && onClear(event);
+  };
+
   render(props = {}) {
     const {
```
Comment (gendeld) [inline src/Input/Input.js:257]: `clearButton` and `onClear` are unused

File: src/Input/index.d.ts [hunk 5]
```diff
@@ -68,10 +68,10 @@ export default class Input extends React.Component<InputProps> {
   static StatusLoading: InputStatusLoading;
   static StatusWarning: InputStatusWarning;
 
-  clear: (event: any) => void;
+  clear: (event?: React.ChangeEvent<HTMLInputElement>) => void;
```
Comment (gendeld) [inline src/Input/index.d.ts:71]: This is triggered by a click on the clear button in `InputSuffix` so the event will be a  `MouseEvent`, I think it's not optional either

File: src/Input/index.d.ts [hunk 6]
```diff
@@ -68,10 +68,10 @@ export default class Input extends React.Component<InputProps> {
...
   blur: () => void;
-  focus: (options: any) => void;
-  handleSuffixOnClear: (event: any) => void;
+  focus: HTMLElement['focus'];
+  handleSuffixOnClear: (event?: React.ChangeEvent<HTMLInputElement>) => void;
```
Comment (gendeld) [inline src/Input/index.d.ts:74]: `MouseEvent` as well

Comment (gendeld) [reply src/Input/index.d.ts:74]: You can delete this

File: src/Input/test/Input.tsx [hunk 7]
```diff
@@ -87,10 +87,10 @@ function InputWithAllProps() {
 function testInstanceMethods() {
   const instance = new Input({});
   instance.blur();
-  instance.clear(null);
-  instance.focus(null);
+  instance.clear();
+  instance.focus({ preventScroll: false });
   instance.select();
-  instance.handleSuffixOnClear(null);
+  instance.handleSuffixOnClear();
```
Comment (gendeld) [inline src/Input/test/Input.tsx:93]: This is an internal method

File: src/Input/index.d.ts (hunk 5 above)
Comment (EgoziE) [reply src/Input/index.d.ts:71]: It's tricky!
All these functions `clear`, `blur`, `focus`, `select` are public and can be triggered manually, therefore you might not get any event.

Comment (EgoziE) [reply src/Input/index.d.ts:71]: Also, I deleted handleSuffixOnClear from here, just noticed it's here by mistake

**Review by gendeld — APPROVED** [review]

**Pull request merged**